core/utils_test.py

```diff
@@ -794,3 +795,144 @@ def test_url_open(self) -> None:
         self.assertEqual(response.getcode(), 200) # type: ignore[attr-defined]
         self.assertEqual(
             response.url, 'http://www.google.com') # type: ignore[attr-defined]
+
+    def test_get_random_int(self) -> None:
+        self.assertLess(utils.get_random_int(5), 5)
+        self.assertGreaterEqual(utils.get_random_int(5), 0)
+        self.assertLess(utils.get_random_int(True), 1)
+        self.assertGreaterEqual(utils.get_random_int(True), 0)
+        with self.assertRaisesRegex( # type: ignore[no-untyped-call]
+            AssertionError, 'Only positive integers allowed'):
+            utils.get_random_int(-1)
+
+    def test_get_random_choice(self) -> None:
+        list_instance = [1, 5, 9, 11, 15]
+        list_instance2: List[str] = []
+        self.assertIn(utils.get_random_choice(list_instance), list_instance)
+        with self.assertRaisesRegex( # type: ignore[no-untyped-call]
+            AssertionError, 'Only non-empty lists allowed'):
+            utils.get_random_choice(list_instance2)
+
+    def test_get_human_readable_time_string(self) -> None:
+        self.assertEqual(
+            'December 12 06:42:12',
+            utils.get_human_readable_time_string(944980932342.38)
+        )
+        with self.assertRaisesRegex( # type: ignore[no-untyped-call]
+            AssertionError, 'Time cannot be negative'):
+            utils.get_human_readable_time_string(-1.42)
+
+    def test_generate_new_session_id(self) -> None:
+        test_string = utils.generate_new_session_id()
+        self.assertEqual(24, len(test_string))
+        self.assertIsInstance(test_string, str)
+        test_string_utf8 = test_string.encode('utf-8')
+        test_string_utf16 = test_string.encode('utf-16')
+        self.assertEqual(
+            len(test_string), len(test_string_utf8))
+        self.assertNotEqual(
+            len(test_string), len(test_string_utf16))
```

Comment:
I don't think this is testing what we want to test. Here's the code we want to test:

https://github.com/oppia/oppia/blob/be72974f7d728b6084764ded43ff74abb836fb83/core/utils.py#L674-L692

`os.urandom` gives us some random bytes, and then `base64.urlsafe_b64encode()` generates a URL-safe, base64 representation of those bytes. Importantly, its output is still of type `bytes`, not of type `str`. We can decode those bytes in different ways to get different strings:

```python
>>> random_bytes = base64.urlsafe_b64encode(os.urandom(10))
>>> random_bytes.decode('utf-8')
'aX6Z9L4wvMubpA=='
>>> random_bytes.decode('utf-16')
'塡娶䰹眴䵶扵䅰㴽'
```

However, for our purposes, we don't really care about whether UTF-8 or UTF-16 (or any other encoding) was used. We only care that the final string is URL-safe base64. For example, the string `'aaaaaaaaaaaaaaaaaaaaaaaa'` is URL-safe base64, so it is a valid session ID. Here we can see the UTF-8 and UTF-16 encodings for this string:

```python
>>> s = 'a' * 24
>>> s
'aaaaaaaaaaaaaaaaaaaaaaaa'
>>> s.encode('utf-8')
b'aaaaaaaaaaaaaaaaaaaaaaaa'
>>> s.encode('utf-16')
b'\xff\xfea\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00'
```

Now consider two possible `generate_new_session_id()` implementations. One returns `b'aaaaaaaaaaaaaaaaaaaaaaaa'.decode('utf-8')` and the other returns `b'\xff\xfea\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00'.decode('utf-16')`. Both implementations will return the same string of 24 `a`s:

```python
>>> b'aaaaaaaaaaaaaaaaaaaaaaaa'.decode('utf-8')
'aaaaaaaaaaaaaaaaaaaaaaaa'
>>> b'\xff\xfea\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00a\x00'.decode('utf-16')
'aaaaaaaaaaaaaaaaaaaaaaaa'
```

Both these implementations generate a 24-character string that is URL-safe and valid base64, so both should pass this test.

The takeaway here is that the test should not (and cannot) test what encoding scheme was used inside `generate_random_string()`. Instead, it should test that the resulting string has the desired properties (24 characters long, base64, URL-safe)

Reply:
From what I have been able to understand online, there is no 100% way to tell if a string is encoded in base64 we can only test if that string can be produced by base64. There are methods that come pretty close but I don't think it will be a good idea to implement that as it will cause failures to occur from time to time.
I have, however, added a test for URL-safe

Reply:
By "valid base64" I just meant that it consists exclusively of characters that are valid in base64. In fact, since we truncate the string, it's probably not truly valid base64. Sorry I wasn't clear

Reply:
So no other changes are required for this test, right?